File: runtime/go/wire_event_context_test.go
package runtime_test

import (
	"context"
	"encoding/json"
	"errors"
	"reflect"
	"testing"
	"time"

	"github.com/Bitspark/nightseam/duplex/go"
	ws "github.com/Bitspark/nightseam/runtime/go"
)

type eventVerifiedKey struct{}
type eventVerifiedPropagator struct{ verified any }

func (p eventVerifiedPropagator) Extract(ctx context.Context, trace ws.Trace) context.Context {
	return context.WithValue(ws.DefaultPropagator.Extract(ctx, trace), eventVerifiedKey{}, p.verified)
}
func (eventVerifiedPropagator) Inject(ctx context.Context) ws.Trace {
	return ws.DefaultPropagator.Inject(ctx)
}

func TestWireEventContextSurvivesPhysicalForwardLocalPairAndMount(t *testing.T) {
	verified := &struct{ source string }{"trusted context"}
	client, server := newPair(t, ws.Options{Propagator: eventVerifiedPropagator{verified}}, ws.Options{})
	access, binding, err := ws.NewWirePair(ws.Options{MaxPendingRequests: 1})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = access.Close(duplex.CodeNormal, "done") })
	stop, err := ws.ForwardWire(server.Wire(), duplex.At(duplex.Mount(map[string]duplex.Wire{"local": access}), []string{"local"}))
	if err != nil {
		t.Fatal(err)
	}
	defer stop()
	model := duplex.At(duplex.Mount(map[string]duplex.Wire{"model": binding}), []string{"model", "events"})
	observed := make(chan context.Context, 1)
	effects := 0
	_, err = ws.RegisterWire(model, []string{"change"}, ws.WireHandlers{Event: func(ctx context.Context, _ json.RawMessage) error {
		observed <- ctx
		if ctx.Value(eventVerifiedKey{}) != verified {
			return errors.New("unverified event")
		}
		effects++
		return nil
	}})
	if err != nil {
		t.Fatal(err)
	}
	meta := map[string]string{"tenant": "explicit", "verified": "cannot manufacture context"}
	if err := ws.EmitWire(ws.WithMeta(context.Background(), meta), client.Wire(), []string{"events", "change"}, nil); err != nil {
		t.Fatal(err)
	}
	ctx := receive(t, observed)
	if ctx.Value(eventVerifiedKey{}) != verified || !reflect.DeepEqual(ws.MetaFrom(ctx), meta) {
		t.Fatalf("lost received context: verified=%v meta=%v", ctx.Value(eventVerifiedKey{}), ws.MetaFrom(ctx))
	}
	_, _ = ws.HandleWire(binding, []string{"barrier"}, func(context.Context, json.RawMessage) (any, error) { return effects, nil })
	var count int
	if err := ws.CallWire(context.Background(), access, []string{"barrier"}, nil, &count); err != nil || count != 1 {
		t.Fatalf("effect count=%d, err=%v", count, err)
	}

	// New outgoing events carry only explicitly supplied metadata. The private
	// received context ends at the next physical boundary.
	returned := make(chan context.Context, 2)
	_, _ = ws.RegisterWire(client.Wire(), []string{"outgoing"}, ws.WireHandlers{Event: func(ctx context.Context, _ json.RawMessage) error { returned <- ctx; return nil }})
	if err := ws.EmitWire(ctx, server.Wire(), []string{"outgoing"}, nil); err != nil {
		t.Fatal(err)
	}
	fresh := receive(t, returned)
	if fresh.Value(eventVerifiedKey{}) != nil || len(ws.MetaFrom(fresh)) != 0 {
		t.Fatalf("ambient context crossed transport: %v %v", fresh.Value(eventVerifiedKey{}), ws.MetaFrom(fresh))
	}
	if err := ws.EmitWire(ws.WithMeta(ctx, ws.MetaFrom(ctx)), server.Wire(), []string{"outgoing"}, nil); err != nil {
		t.Fatal(err)
	}
	if got := ws.MetaFrom(receive(t, returned)); !reflect.DeepEqual(got, meta) {
		t.Fatalf("explicit outgoing metadata=%v", got)
	}
	_ = server.Close()
	select {
	case <-ctx.Done():
	case <-time.After(time.Second):
		t.Fatal("event context lost its physical connection lifetime")
	}
}

func TestWireEventMetadataCannotSupplyVerifiedContext(t *testing.T) {
	client, server := newPair(t, ws.Options{}, ws.Options{})
	access, binding, err := ws.NewWirePair(ws.Options{})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = access.Close(duplex.CodeNormal, "done") })
	stop, err := ws.ForwardWire(server.Wire(), access)
	if err != nil {
		t.Fatal(err)
	}
	defer stop()
	denied := make(chan bool, 1)
	_, _ = ws.RegisterWire(binding, []string{"guard"}, ws.WireHandlers{Event: func(ctx context.Context, _ json.RawMessage) error {
		if ctx.Value(eventVerifiedKey{}) == nil {
			denied <- true
			return errors.New("event denied")
		}
		denied <- false
		return nil
	}})
	if err := ws.EmitWire(ws.WithMeta(context.Background(), map[string]string{"verified": "yes"}), client.Wire(), []string{"guard"}, nil); err != nil {
		t.Fatal(err)
	}
	if !receive(t, denied) {
		t.Fatal("metadata bypassed the event guard")
	}
}

func TestWireEventContextStopsAtAnotherPhysicalBoundary(t *testing.T) {
	verified := &struct{}{}
	client, incoming := newPair(t, ws.Options{Propagator: eventVerifiedPropagator{verified}}, ws.Options{})
	outgoing, server := newPair(t, ws.Options{}, ws.Options{})
	stop, err := ws.ForwardWire(incoming.Wire(), outgoing.Wire())
	if err != nil {
		t.Fatal(err)
	}
	defer stop()
	observed := make(chan context.Context, 1)
	_, _ = ws.RegisterWire(server.Wire(), []string{"event"}, ws.WireHandlers{Event: func(ctx context.Context, _ json.RawMessage) error { observed <- ctx; return nil }})
	if err := ws.EmitWire(ws.WithMeta(context.Background(), map[string]string{"explicit": "yes"}), client.Wire(), []string{"event"}, nil); err != nil {
		t.Fatal(err)
	}
	ctx := receive(t, observed)
	if ctx.Value(eventVerifiedKey{}) != nil || ws.MetaFrom(ctx)["explicit"] != "yes" {
		t.Fatalf("physical boundary: private=%v meta=%v", ctx.Value(eventVerifiedKey{}), ws.MetaFrom(ctx))
	}
}

func TestWireLocalEventsUseSuppliedPropagator(t *testing.T) {
	verified := &struct{}{}
	a, b, err := ws.NewWirePair(ws.Options{Propagator: eventVerifiedPropagator{verified}})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = a.Close(duplex.CodeNormal, "done") })
	observed := make(chan context.Context, 1)
	_, _ = ws.RegisterWire(b, []string{"event"}, ws.WireHandlers{Event: func(ctx context.Context, _ json.RawMessage) error { observed <- ctx; return nil }})
	if err := ws.EmitWire(context.Background(), a, []string{"event"}, nil); err != nil {
		t.Fatal(err)
	}
	if receive(t, observed).Value(eventVerifiedKey{}) != verified {
		t.Fatal("local event bypassed configured propagator")
	}
}

File: runtime/ts/src/wire-event-context.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { at, mount, pipe, type Message, type Wire } from '@nightseam/duplex';
import { DuplexPeer } from './peer.ts';
import type { PeerOptions } from './peer.ts';
import { emitWire, forwardWire, registerWire, callWire, handleWire } from './wire.ts';
import type { WireEventContext } from './wire.ts';
import { wirePair } from './wire-pair.ts';
import { defaultPropagator } from './trace.ts';

function deferred<T>() {
  let resolve!: (value: T) => void;
  const promise = new Promise<T>((yes) => {
    resolve = yes;
  });
  return { promise, resolve };
}
async function physical(serverOptions: PeerOptions = {}) {
  const [a, b] = pipe();
  const client = new DuplexPeer(),
    server = new DuplexPeer({ ...serverOptions, role: 'server' });
  await Promise.all([client.attach(a), server.attach(b)]);
  return {
    client,
    server,
    close: () => {
      client.close();
      server.close();
    },
  };
}

test('physical events keep verified context through forwarding local pair and mount without ambient outgoing metadata', async (t) => {
  const verified = Object.freeze({ source: 'trusted context' });
  const marker = Symbol('verified');
  const peers = await physical({
    propagator: {
      extract: (context, trace) => {
        defaultPropagator.extract(context, trace);
        Object.defineProperty(context, marker, { value: verified });
      },
      inject: defaultPropagator.inject,
    },
  });
  t.after(peers.close);
  const [access, binding] = wirePair({ maxPendingRequests: 1 });
  t.after(() => access.close());
  t.after(forwardWire(peers.server.wire(), at(mount(new Map([['local', access]])), ['local'])));
  const model = at(mount(new Map([['model', binding]])), ['model', 'events']);
  const observed = deferred<WireEventContext>();
  let effects = 0;
  registerWire(model, ['change'], {
    event: (_data, context) => {
      observed.resolve(context);
      if ((context as unknown as Record<symbol, unknown>)[marker] !== verified) throw new Error('Unverified event');
      effects++;
    },
  });
  const meta = { tenant: 'explicit', verified: 'cannot manufacture context' };
  emitWire(peers.client.wire(), ['events', 'change'], null, { meta });
  const context = await observed.promise;
  assert.equal((context as unknown as Record<symbol, unknown>)[marker], verified);
  assert.deepEqual(context.meta, meta);
  assert.equal((context as unknown as { peer: DuplexPeer }).peer, peers.server);
  handleWire(binding, ['barrier'], () => effects);
  assert.equal(await callWire(access, ['barrier']), 1);
  const received: WireEventContext[] = [];
  const arrived = deferred<void>();
  registerWire(peers.client.wire(), ['outgoing'], {
    event: (_value, context) => {
      received.push(context);
      if (received.length === 2) arrived.resolve();
    },
  });
  emitWire(peers.server.wire(), ['outgoing'], null, { context });
  emitWire(peers.server.wire(), ['outgoing'], null, { context, meta: context.meta });
  await arrived.promise;
  assert.equal((received[0] as unknown as Record<symbol, unknown>)[marker], undefined);
  assert.equal(received[0]!.meta, undefined);
  assert.deepEqual(received[1]!.meta, meta);
});

test('event metadata cannot create the private context required by an effect guard', async (t) => {
  const marker = Symbol('verified');
  const peers = await physical();
  t.after(peers.close);
  const [access, binding] = wirePair();
  t.after(() => access.close());
  t.after(forwardWire(peers.server.wire(), access));
  const denied = deferred<boolean>();
  registerWire(binding, ['guard'], {
    event: (_value, context) => {
      if (!(context as unknown as Record<symbol, unknown>)[marker]) {
        denied.resolve(true);
        throw new Error('Denied');
      }
      denied.resolve(false);
    },
  });
  emitWire(peers.client.wire(), ['guard'], null, { meta: { verified: 'yes' } });
  assert.equal(await denied.promise, true);
});

test('a forwarded event context ends at the next physical transport boundary', async (t) => {
  const marker = Symbol('verified');
  const source = await physical({
    propagator: {
      extract: (context, trace) => {
        defaultPropagator.extract(context, trace);
        Object.defineProperty(context, marker, { value: true });
      },
      inject: defaultPropagator.inject,
    },
  });
  const destination = await physical();
  t.after(source.close);
  t.after(destination.close);
  t.after(forwardWire(source.server.wire(), destination.client.wire()));
  const observed = deferred<WireEventContext>();
  registerWire(destination.server.wire(), ['event'], { event: (_value, context) => observed.resolve(context) });
  emitWire(source.client.wire(), ['event'], null, { meta: { explicit: 'yes' } });
  const context = await observed.promise;
  assert.equal((context as unknown as Record<symbol, unknown>)[marker], undefined);
  assert.deepEqual(context.meta, { explicit: 'yes' });
});

test('pure local events use the configured propagator', async (t) => {
  const marker = Symbol('local');
  const [a, b] = wirePair({
    propagator: {
      extract: (context, trace) => {
        defaultPropagator.extract(context, trace);
        Object.defineProperty(context, marker, { value: true });
      },
      inject: defaultPropagator.inject,
    },
  });
  t.after(() => a.close());
  const observed = deferred<WireEventContext>();
  registerWire(b, ['event'], { event: (_value, context) => observed.resolve(context) });
  emitWire(a, ['event']);
  assert.equal(((await observed.promise) as unknown as Record<symbol, unknown>)[marker], true);
});

test('a custom event propagator cannot rewrite the received frame when it is forwarded', async (t) => {
  const changed = { traceparent: '00-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-bbbbbbbbbbbbbbbb-01' };
  const peers = await physical({
    propagator: {
      extract: (context) => {
        context.trace = changed;
      },
      inject: defaultPropagator.inject,
    },
  });
  t.after(peers.close);
  const [access, binding] = wirePair();
  t.after(() => access.close());
  const sent = deferred<Message>(),
    received = deferred<Message>();
  const source: Wire = {
    ...peers.client.wire(),
    send: (path, message) => {
      sent.resolve(message);
      peers.client.wire().send(path, message);
    },
  };
  const destination: Wire = {
    ...access,
    send: (path, message) => {
      received.resolve(message);
      access.send(path, message);
    },
  };
  t.after(forwardWire(peers.server.wire(), destination));
  const delivered = deferred<WireEventContext>();
  registerWire(binding, ['event'], { event: (_value, context) => delivered.resolve(context) });
  emitWire(source, ['event']);
  assert.equal((await received.promise).frame.traceparent, (await sent.promise).frame.traceparent);
  assert.equal((await delivered.promise).trace, changed);
});
